### PR TITLE
Dedicated title/description input fields

### PR DESCRIPTION
**Pull request by tuancoltech**
* Support dedicated title/description input fields.
https://github.com/ARK-Builders/ARK-Memo/issues/22

    (`Titling rule:` https://github.com/ARK-Builders/ARK-Memo/issues/22#issuecomment-1864432794)

* Fix a minor UI issue with the `Add note` FAB.

### CODE REVIEW

File: app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditGraphicNotesFragment.kt [hunk 1]
```diff
@@ -77,22 +69,25 @@ class EditGraphicNotesFragment: Fragment(R.layout.fragment_edit_notes) {
...
+        noteTitle.hint = getString(R.string.hint_new_graphical_note)
         noteTitle.setText(title)
         noteTitle.addTextChangedListener(noteTitleChangeListener)
         notesCanvas.isVisible = true
         notesCanvas.setViewModel(graphicNotesViewModel)
         saveButton.setOnClickListener {
             val svg = graphicNotesViewModel.svg()
             val note = GraphicNote(
-                title = title.ifEmpty { defaultTitle },
+                title = binding.noteTitle.text.toString(),
```
Comment (ShubertMunthali) [inline app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditGraphicNotesFragment.kt:80]: variable ```title``` should be used here

File: app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditTextNotesFragment.kt [hunk 2]
```diff
@@ -74,6 +63,9 @@ class EditTextNotesFragment: Fragment(R.layout.fragment_edit_notes) {
 
             override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                 title = s?.toString() ?: ""
+                if (title.isEmpty()) {
+                    binding.noteTitle.hint = getString(R.string.hint_new_text_note)
```
Comment (ShubertMunthali) [inline app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditTextNotesFragment.kt:67]: I think its better to have timestamp on the default tilte to make it easy for user to identify a specfic note among notes saved with default titles. What do you think @kirillt

**Review by ShubertMunthali — COMMENTED** [review]
It would also be better to allow user save title and description separately, right now title and description are tightly coupled to content despite having separate storage #39

File: app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditGraphicNotesFragment.kt (hunk 1 above)
Comment (tuancoltech) [reply app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditGraphicNotesFragment.kt:80]: @ShubertMunthali Can you elaborate further?
`binding.noteTitle.text.toString()` is basically the same as `title` unless user enters his own title.

File: app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditTextNotesFragment.kt (hunk 2 above)
Comment (tuancoltech) [reply app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditTextNotesFragment.kt:67]: @ShubertMunthali We're still showing note's timestamp which is taken from `modified` property of the note's resource field.

Screenshot 2024-01-09 at 23 03 06

Comment (kirillt) [reply app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditTextNotesFragment.kt:67]: @ShubertMunthali I think we don't need the timestamp in the title, but we should have it somewhere around.

@tuancoltech looks good, but this timestamp can be updated externally. We need https://github.com/ARK-Builders/arklib-android/issues/135 for proper UX (so the timestamp would stick to the resource forever).

**Comment by kirillt** [comment]
@ShubertMunthali not sure what do you mean, I see that the title is stored into new `.ark/user/properties` storage as JSON. It seems to allow empty description and/or empty title.

@tuancoltech the feature works good, the only issue is that it's not possible to modify title after initial creation if the content isn't changed. Because `[save]` button doesn't allow saving same note twice. It shouldn't update the content, but any properties updates should be saved.

**Comment by tuancoltech** [comment]
> @ShubertMunthali not sure what do you mean, I see that the title is stored into new `.ark/user/properties` storage as JSON. It seems to allow empty description and/or empty title.
> 
> @tuancoltech the feature works good, the only issue is that it's not possible to modify title after initial creation if the content isn't changed. Because `[save]` button doesn't allow saving same note twice. It shouldn't update the content, but any properties updates should be saved.

@kirillt At the moment, if we modify a note's content, a new note will be created instead of overwriting the existing note file.
Do we want to duplicate the note files, or overwriting in the same file in case user modify the note's `description` and/or `title`?

**Comment by ShubertMunthali** [comment]
> @ShubertMunthali not sure what do you mean, I see that the title is stored into new `.ark/user/properties` storage as JSON. It seems to allow empty description and/or empty title.
> 
> @tuancoltech the feature works good, the only issue is that it's not possible to modify title after initial creation if the content isn't changed. Because `[save]` button doesn't allow saving same note twice. It shouldn't update the content, but any properties updates should be saved.

@kirillt I mean what you just explained, that save button does not work if only properties are edited, it works when content is edited. we need it to work with properties edited alone, or with properties edited together with content. That's what is required in #39

**Comment by ShubertMunthali** [comment]
> > @ShubertMunthali not sure what do you mean, I see that the title is stored into new `.ark/user/properties` storage as JSON. It seems to allow empty description and/or empty title.
> > @tuancoltech the feature works good, the only issue is that it's not possible to modify title after initial creation if the content isn't changed. Because `[save]` button doesn't allow saving same note twice. It shouldn't update the content, but any properties updates should be saved.
> 
> @kirillt At the moment, if we modify a note's content, a new note will be created instead of overwriting the existing note file. Do we want to duplicate the note files, or overwriting in the same file in case user modify the note's `description` and/or `title`?

If only properties change we should overwrite, but whenever content changes we need a new file. Version tracking in #13 will handle the duplicates properly.

File: app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditGraphicNotesFragment.kt (hunk 1 above)
Comment (ShubertMunthali) [reply app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditGraphicNotesFragment.kt:80]: It's a little confusing where the title is coming from :). My first glance at that line I thought the title is missing :)

**Comment by kirillt** [comment]
Yep, @ShubertMunthali is right here. We're missing ability to edit title/description when the content is untouched. We don't need to save the ckntent twice in this case, only update `user/properties` storage for the current resource id.

File: app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditTextNotesFragment.kt (hunk 2 above)
Comment (tuancoltech) [reply app/src/main/java/dev/arkbuilders/arkmemo/ui/fragments/EditTextNotesFragment.kt:67]: @kirillt Sure, let's get back to this after we have `created` date thing ready.

**Comment by tuancoltech** [comment]
> Yep, @ShubertMunthali is right here. We're missing ability to edit title/description when the content is untouched. We don't need to save the content twice in this case, only update `user/properties` storage for the current resource id.

@ShubertMunthali @kirillt Thanks for the idea, guys. I'm making changes to align with this.

**Comment by tuancoltech** [comment]
@ShubertMunthali @kirillt Please help reviewing again.
Title/description are now validated and stored separately from content.

**Comment by kirillt** [comment]
@tuancoltech Good, but couple of nitpicks:
- [ ] If we create a note without providing title or description, the empty JSON is written to the properties storage: `{"titles":[""],"descriptions":[""]}`. Can we avoid it in the app by simply treating absent properties file as empty JSON?
  * If we need to fix `arklib-android` for this, then let's create the issue there.
  * If we can fix it only in the app code, let's do it.
  
- [ ] Graphical notes are modified in weird way when their title or description is modified.

Steps:
1. Create new graphical note and save.
2. Open it and change the title, save.
3. Open it again and change the description, save.

Result: 3 different SVG files.
```
$ xmllint --format 2035-514095390.svg > /tmp/a
$ xmllint --format 2035-1319256942.svg > /tmp/b
$ xmllint --format 2035-2093532239.svg > /tmp/c
$ diff /tmp/{a,b}
2c2
< <svg xmlns="http://www.w3.org/2000/svg" viewBox="0.0 0.0 1440.0 2550.0">
---
> <svg xmlns="http://www.w3.org/2000/svg" viewBox="0.0 0.0 1440.0 1535.0">
[kirill@lenovo Notes]$ diff /tmp/{b,c}
2c2
< <svg xmlns="http://www.w3.org/2000/svg" viewBox="0.0 0.0 1440.0 1535.0">
---
> <svg xmlns="http://www.w3.org/2000/svg" viewBox="0.0 0.0 1440.0 1398.0">
```

For some reason, `viewBox` property is updated.

**Comment by tuancoltech** [comment]
> * If we create a note without providing title or description, the empty JSON is written to the properties storage: `{"titles":[""],"descriptions":[""]}`

@kirillt For the empty `JSON` issue, we could only handle that from `arklib-android`.
I created an issue for it here: https://github.com/ARK-Builders/arklib-android/issues/137

For the second issue, it's because we are updating the `viewBox` of current `SVG` whenever there's size change of the drawing view as per in below screenshot.
I'm clarifying this with @ShubertMunthali .

Screenshot 2024-01-16 at 12 40 04

**Comment by kirillt** [comment]
Thank you @tuancoltech, that's perfect.

**Comment by kirillt** [comment]
So, I guess it's easy to fix as soon as @ShubertMunthali share some info and we'll merge then.

**Comment by ShubertMunthali** [comment]
We need to set dimensions for newly created note only here. @tuancoltech that's a bug indeed, we need a fix. You can provide it, if yo don't mind

**Comment by tuancoltech** [comment]
> We need to set dimensions for newly created note only here. @tuancoltech that's a bug indeed, we need a fix. You can provide it, if yo don't mind

@ShubertMunthali Definitely, I have fixed it by removing the unnecessary `onSizeChanged` method.
Please re-review @kirillt !

**Review by kirillt — APPROVED** [review]
Looks good, thanks!

**Pull request merged**